File: backend/app/services/infrastructure/circuit_breaker.py
"""
Circuit Breaker implementation for backend services
Provides fault tolerance for database operations and external API calls
"""

import asyncio
import logging
import time
from contextlib import asynccontextmanager
from dataclasses import dataclass
from enum import Enum
from typing import Any

logger = logging.getLogger(__name__)


class CircuitBreakerState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


@dataclass
class CircuitBreakerConfig:
    failure_threshold: int = 5  # Number of failures before opening
    recovery_timeout: float = 60.0  # Seconds to wait before trying again
    expected_exception: tuple = (Exception,)  # Exception types to catch
    success_threshold: int = 3  # Successes needed in half-open state
    timeout: float = 30.0  # Operation timeout


class CircuitBreaker:
    """Circuit breaker implementation for fault tolerance"""

    def __init__(self, name: str, config: CircuitBreakerConfig = None):
        self.name = name
        self.config = config or CircuitBreakerConfig()

        self.state = CircuitBreakerState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time: float | None = None

        self._lock = asyncio.Lock()

    async def _should_attempt_reset(self) -> bool:
        """Check if enough time has passed to attempt reset"""
        if self.last_failure_time is None:
            return False
        return time.time() - self.last_failure_time >= self.config.recovery_timeout

    async def _record_success(self):
        """Record a successful operation"""
        async with self._lock:
            if self.state == CircuitBreakerState.HALF_OPEN:
                self.success_count += 1
                if self.success_count >= self.config.success_threshold:
                    self._reset()
            elif self.state == CircuitBreakerState.CLOSED:
                self.failure_count = max(0, self.failure_count - 1)  # Decay failures

    async def _record_failure(self):
        """Record a failed operation"""
        async with self._lock:
            self.failure_count += 1
            self.last_failure_time = time.time()

            if self.failure_count >= self.config.failure_threshold:
                self.state = CircuitBreakerState.OPEN
                logger.warning(
                    f"Circuit breaker '{self.name}' opened after {self.failure_count} failures"
                )

    def _reset(self):
        """Reset circuit breaker to closed state"""
        self.state = CircuitBreakerState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time = None
        logger.info(f"Circuit breaker '{self.name}' reset to closed state")

    def _can_attempt(self) -> bool:
        """Check if operation can be attempted - sync version"""
        if self.state == CircuitBreakerState.CLOSED:
            return True
        elif self.state == CircuitBreakerState.OPEN:
            # For sync check, we use current time without awaiting
            if self.last_failure_time is None:
                return False
            return time.time() - self.last_failure_time >= self.config.recovery_timeout
        else:  # HALF_OPEN
            return True

    def attempt_operation_sync(self):
        """Sync context manager for circuit breaker protected operations"""
        from contextlib import contextmanager

        @contextmanager
        def sync_context():
            if not self._can_attempt():
                raise CircuitBreakerOpenException(
                    f"Circuit breaker '{self.name}' is open"
                )

            if self.state == CircuitBreakerState.OPEN:
                self.state = CircuitBreakerState.HALF_OPEN
                logger.info(f"Circuit breaker '{self.name}' entering half-open state")

            try:
                yield
                # Record success in a way that doesn't block (or just skip for sync if risk is low)
                # For SQLite, the risk is low. For 99.99% uptime, we should ideally use a threadsafe non-async counter.
                if self.state == CircuitBreakerState.HALF_OPEN:
                    self.success_count += 1
                    if self.success_count >= self.config.success_threshold:
                        self._reset()
                elif self.state == CircuitBreakerState.CLOSED:
                    self.failure_count = max(0, self.failure_count - 1)
            except self.config.expected_exception as e:
                self.failure_count += 1
                self.last_failure_time = time.time()
                if self.failure_count >= self.config.failure_threshold:
                    self.state = CircuitBreakerState.OPEN
                    logger.warning(
                        f"Circuit breaker '{self.name}' opened after {self.failure_count} failures"
                    )
                raise e

        return sync_context()

    @asynccontextmanager
    async def attempt_operation(self):
        """Context manager for circuit breaker protected operations"""
        if not self._can_attempt():
            raise CircuitBreakerOpenException(f"Circuit breaker '{self.name}' is open")

        if self.state == CircuitBreakerState.OPEN:
            self.state = CircuitBreakerState.HALF_OPEN
            logger.info(f"Circuit breaker '{self.name}' entering half-open state")

        try:
            yield
            await self._record_success()
        except self.config.expected_exception as e:
            await self._record_failure()
            raise e

    def get_status(self) -> dict[str, Any]:
        """Get circuit breaker status"""
        return {
            "name": self.name,
            "state": self.state.value,
            "failure_count": self.failure_count,
            "success_count": self.success_count,
            "last_failure_time": self.last_failure_time,
            "config": {
                "failure_threshold": self.config.failure_threshold,
                "recovery_timeout": self.config.recovery_timeout,
                "success_threshold": self.config.success_threshold,
                "timeout": self.config.timeout,
            },
        }


class CircuitBreakerOpenException(Exception):
    """Exception raised when circuit breaker is open"""


# Global circuit breaker registry
_circuit_breakers: dict[str, CircuitBreaker] = {}


def get_circuit_breaker(
    name: str, config: CircuitBreakerConfig = None
) -> CircuitBreaker:
    """Get or create a circuit breaker instance"""
    if name not in _circuit_breakers:
        _circuit_breakers[name] = CircuitBreaker(name, config)
    return _circuit_breakers[name]


def get_all_circuit_breakers() -> dict[str, dict[str, Any]]:
    """Get status of all circuit breakers"""
    return {name: cb.get_status() for name, cb in _circuit_breakers.items()}


# Decorator for circuit breaker protection
def circuit_breaker(name: str, config: CircuitBreakerConfig = None):
    """Decorator to apply circuit breaker protection to both sync and async functions"""
    cb = get_circuit_breaker(name, config)
    import functools
    import inspect

    def decorator(func):
        if inspect.iscoroutinefunction(func):

            @functools.wraps(func)
            async def async_wrapper(*args, **kwargs):
                async with cb.attempt_operation():
                    return await func(*args, **kwargs)

            return async_wrapper
        else:

            @functools.wraps(func)
            def sync_wrapper(*args, **kwargs):
                with cb.attempt_operation_sync():
                    return func(*args, **kwargs)

            return sync_wrapper

    return decorator

File: frontend/src/lib/circuitBreaker.ts
import { secureLogger } from "@/utils/secureLogger";

export enum CircuitState {
  CLOSED = "CLOSED", // Normal operation
  OPEN = "OPEN", // Circuit is open, failing fast
  HALF_OPEN = "HALF_OPEN", // Testing if service recovered
}

export interface CircuitBreakerConfig {
  failureThreshold: number; // Number of failures before opening circuit
  recoveryTimeout: number; // Time in ms before attempting recovery
  monitoringPeriod: number; // Time window in ms to count failures
  successThreshold: number; // Number of successes needed in HALF_OPEN to close
}

export interface CircuitBreakerStats {
  state: CircuitState;
  failureCount: number;
  successCount: number;
  lastFailureTime: number | null;
  lastSuccessTime: number | null;
  totalRequests: number;
  totalFailures: number;
  totalSuccesses: number;
}

class CircuitBreaker {
  private state: CircuitState = CircuitState.CLOSED;
  private failureCount: number = 0;
  private successCount: number = 0;
  private lastFailureTime: number | null = null;
  private lastSuccessTime: number | null = null;
  private totalRequests: number = 0;
  private totalFailures: number = 0;
  private totalSuccesses: number = 0;

  constructor(
    private name: string,
    private config: CircuitBreakerConfig,
  ) {}

  async execute<T>(operation: () => Promise<T>): Promise<T> {
    this.totalRequests++;

    // Check if circuit should transition from OPEN to HALF_OPEN
    if (this.state === CircuitState.OPEN) {
      if (this.shouldAttemptRecovery()) {
        this.state = CircuitState.HALF_OPEN;
        this.successCount = 0;
        secureLogger.info(
          "Circuit Breaker",
          `Circuit breaker ${this.name} transitioning to HALF_OPEN`,
        );
      } else {
        throw new Error(`Circuit breaker ${this.name} is OPEN - failing fast`);
      }
    }

    try {
      const result = await operation();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  private onSuccess(): void {
    this.totalSuccesses++;
    this.lastSuccessTime = Date.now();

    if (this.state === CircuitState.HALF_OPEN) {
      this.successCount++;
      if (this.successCount >= this.config.successThreshold) {
        this.reset();
        secureLogger.info(
          "Circuit Breaker",
          `Circuit breaker ${this.name} closed after successful recovery`,
        );
      }
    } else if (this.state === CircuitState.CLOSED) {
      // Reset failure count on success in closed state
      this.failureCount = 0;
    }
  }

  private onFailure(): void {
    this.totalFailures++;
    this.failureCount++;
    this.lastFailureTime = Date.now();

    if (this.state === CircuitState.HALF_OPEN) {
      // Failed during recovery test, go back to OPEN
      this.state = CircuitState.OPEN;
      this.successCount = 0;
      secureLogger.warn(
        "Circuit Breaker",
        `Circuit breaker ${this.name} recovery failed, returning to OPEN`,
      );
    } else if (
      this.state === CircuitState.CLOSED &&
      this.failureCount >= this.config.failureThreshold
    ) {
      this.state = CircuitState.OPEN;
      secureLogger.warn(
        "Circuit Breaker",
        `Circuit breaker ${this.name} opened due to ${this.failureCount} failures`,
      );
    }
  }

  private shouldAttemptRecovery(): boolean {
    if (!this.lastFailureTime) return false;
    return Date.now() - this.lastFailureTime >= this.config.recoveryTimeout;
  }

  private reset(): void {
    this.state = CircuitState.CLOSED;
    this.failureCount = 0;
    this.successCount = 0;
    this.lastFailureTime = null;
  }

  getStats(): CircuitBreakerStats {
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      lastFailureTime: this.lastFailureTime,
      lastSuccessTime: this.lastSuccessTime,
      totalRequests: this.totalRequests,
      totalFailures: this.totalFailures,
      totalSuccesses: this.totalSuccesses,
    };
  }

  // Manual control methods
  forceOpen(): void {
    this.state = CircuitState.OPEN;
    secureLogger.warn(
      "Circuit Breaker",
      `Circuit breaker ${this.name} manually opened`,
    );
  }

  forceClose(): void {
    this.reset();
    secureLogger.info(
      "Circuit Breaker",
      `Circuit breaker ${this.name} manually closed`,
    );
  }

  forceHalfOpen(): void {
    this.state = CircuitState.HALF_OPEN;
    this.successCount = 0;
    secureLogger.info(
      "Circuit Breaker",
      `Circuit breaker ${this.name} manually set to HALF_OPEN`,
    );
  }
}

// Circuit Breaker Registry
class CircuitBreakerRegistry {
  private breakers = new Map<string, CircuitBreaker>();

  create(name: string, config: CircuitBreakerConfig): CircuitBreaker {
    if (this.breakers.has(name)) {
      return this.breakers.get(name)!;
    }

    const breaker = new CircuitBreaker(name, config);
    this.breakers.set(name, breaker);
    return breaker;
  }

  get(name: string): CircuitBreaker | undefined {
    return this.breakers.get(name);
  }

  getAll(): Map<string, CircuitBreaker> {
    return new Map(this.breakers);
  }

  getStats(): Record<string, CircuitBreakerStats> {
    const stats: Record<string, CircuitBreakerStats> = {};
    for (const [name, breaker] of this.breakers) {
      stats[name] = breaker.getStats();
    }
    return stats;
  }
}

// Default configurations
export const DEFAULT_CIRCUIT_CONFIGS = {
  api: {
    failureThreshold: 5,
    recoveryTimeout: 60000, // 1 minute
    monitoringPeriod: 300000, // 5 minutes
    successThreshold: 3,
  },
  database: {
    failureThreshold: 3,
    recoveryTimeout: 30000, // 30 seconds
    monitoringPeriod: 180000, // 3 minutes
    successThreshold: 2,
  },
  externalService: {
    failureThreshold: 3,
    recoveryTimeout: 120000, // 2 minutes
    monitoringPeriod: 600000, // 10 minutes
    successThreshold: 2,
  },
} as const;

// Global registry instance
export const circuitBreakerRegistry = new CircuitBreakerRegistry();

// Convenience functions
export function createCircuitBreaker(
  name: string,
  config: CircuitBreakerConfig,
): CircuitBreaker {
  return circuitBreakerRegistry.create(name, config);
}

export function getCircuitBreaker(name: string): CircuitBreaker | undefined {
  return circuitBreakerRegistry.get(name);
}

export function getAllCircuitBreakerStats(): Record<
  string,
  CircuitBreakerStats
> {
  return circuitBreakerRegistry.getStats();
}
